ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,357 @@
+  if TableSet.mem tblname (Database.tableset db) then begin
+    if (Table.mem objref (TableSet.find tblname (Database.tableset db))) then begin
+      if (Row.mem fldname (Table.find objref (TableSet.find tblname (Database.tableset db)))) then
+        apply stat.modified tblname objref fldname newval db
+      else
+        begin
+          if stat.created = stat.modified then
+            apply stat.created tblname objref fldname newval db
+          else
+            apply stat.modified tblname objref fldname newval (apply stat.created tblname objref fldname (Schema.Value.String "XXX") db)
+        end
+    end
+    else
+      begin
+        if stat.created = stat.modified then
+          apply stat.created tblname objref fldname newval db
+        else
+          apply stat.modified tblname objref fldname newval (apply stat.created tblname objref fldname (Schema.Value.String (tblname^objref)) db)
+      end
+  end
+  else
+    apply stat.modified tblname objref fldname newval (make_empty_table {stat with created=(-1L);} tblname db)
+
+
+let compare_by_tblname (c1: (string * int)) (c2: (string * int)) =
+  String.compare (fst c1) (fst c2)
+
+let count_check (c1_list: (string * int) list) (c2_list: (string * int) list) =
+  let c1_list = List.sort compare_by_tblname c1_list in
+  let c2_list = List.sort compare_by_tblname c2_list in
+  try
+    List.for_all2 (fun (a,b) (c,d) -> a=c && b=d) c1_list c2_list
+  with invalid_arg -> false
+
+let compare_by_modified u1 u2 =
+  if u1.stat.modified = u2.stat.modified then
+    Int64.compare u1.stat.created u2.stat.created
+  else
+    Int64.compare u1.stat.modified u2.stat.modified
+
+let f_field u (field:field) =
+  {u with
+   stat = field.stat;
+   fldname = field.fldname;
+   value = field.value;}
+
+let f_row u (row:row) =
+  let u = {u with
+           stat = {row.stat with modified = (max row.stat.modified u.stat.modified)};
+           objref = row.objref;} in
+  match row.fields with
+  | [] -> [u]
+  | _ -> List.rev_map (f_field u) row.fields
+
+let f_table (table:table) =
+  let u = {stat = table.stat;
+           tblname = table.tblname;
+           objref = "";
+           fldname = "";
+           value = (Schema.Value.String "");
+          } in
+  match table.rows with
+  | [] -> [u]
+  | _ -> List.concat (List.rev_map (f_row u) table.rows)
+
+let f_all table_list =
+  List.sort compare_by_modified (List.concat (List.rev_map f_table table_list))
+
+
+let make_change_with_db_reply db (update:update) =
+  match update.objref with
+  | "" -> make_empty_table update.stat update.tblname db
+  | _ -> (match update.fldname with
+      | "" -> make_from_barrier update.stat update.tblname update.objref db
+      | _ -> make_change update.stat update.tblname update.objref update.fldname update.value db)
+
+let make_delete_with_db_reply db delete =
+  remove_row delete.table delete.key db
+
+let modified_check db tblname =
+  let _t = TableSet.find tblname (Database.tableset db) in
```

Comment:
An underscore typically indicates an unused value. To indicate a derived value; `t'` (prime) could be used.